engine/Real.cs

```diff
@@ -0,0 +1,456 @@
+		public bool IsAbsolutelyNear(Real o, Real slop)
+		{
+			if (slop < 0.0)
+			{
+				throw new ArgumentException("slop cannot be negative.","slop");
+			}
+			bool? sc = NearSpecialCases(o);
+			if (sc != null)
+			{
+				return sc.Value;
+			}
+
+			Real diff = this - o;
+			if (diff < 0.0) diff = -diff;
+			return diff <= slop;
+		}
+
+		/// <summary>
+		/// Compares this Real value for being near another one, with a proportional difference no greater than a
+		/// certain limit. NaN is not near anything (including NaN), infinities are only near the same infinity,
+		/// and zero is only near zero (negative and positive zero are indistinct). Values of different signs are
+		/// not relatively near each other. Proportion is relative to the smaller value.
+		/// </summary>
+		/// <param name="o">Value to compare this to.</param>
+		/// <param name="slop">Proportional difference tolerated between values. Nonnegative.</param>
+		/// <returns>Whether this is near o, within the given proportion.</returns>
+		public bool IsRelativelyNear(Real o, Real slop)
+		{
+			if (slop < 0.0)
+			{
+				throw new ArgumentException("slop cannot be negative.", "slop");
+			}
+			bool? sc = NearSpecialCases(o);
+			if (sc != null)
+			{
+				return sc.Value;
+			}
+
+			if (o == 0.0)
+			{
+				// Zero is not relatively near anything else.
+				return this == 0.0;
+			}
+
+			Real big, little;
+			if (this < 0.0)
+			{
+				if (o > 0.0) return false;
+				// Negative values; "big" and "little" refer to magnitude and the signs will cancel
+				// out during division, so reverse the obvious semantics for which goes where.
+				if (this > o)
+				{
+					big = o;
+					little = this;
+				}
+				else
+				{
+					big = this;
+					little = o;
+				}
+			}
+			else
+			{
+				if (o < 0.0) return false;
+				if (this > o)
+				{
+					big = this;
+					little = o;
+				}
+				else
+				{
+					big = o;
+					little = this;
+				}
+			}
+
+			Real ratio = big / little;
+			return (ratio - 1.0) <= slop;
+		}
+
+		/// <summary>
+		/// Returns whether this is near the other value, within either an absolute or relative
+		/// range. NaN values are never near anything. Infinite values are only near themselves.
+		/// Otherwise, values are near each other if the absolute value of the difference between
+		/// them is less than the slop value, or if the proportional difference between them is
+		/// less than the slop value (calculated relative to the smaller value).
+		/// </summary>
+		/// <param name="o">Value to compare against.</param>
+		/// <param name="slop">Allowable amount of error, either absolute or proportional, for
+		/// values to be considered near each other. Nonnegative.</param>
+		/// <returns></returns>
+		public bool IsNear(Real o, Real slop) => IsAbsolutelyNear(o, slop) || IsRelativelyNear(o, slop);
+#endregion
+
+#region Unary operators
+		// Note: NaN is neither true nor false.
+		public static bool operator true(Real r)
+		{
+#if !REALTYPE_DECIMAL
```

Comment:
Here you imply all non-decimal types have NaN defined. This is indeed currently the case, but in the future we might have that there are additional types that are added which would also not have NaN defined, so then using
```suggestion
#if (REALTYPE_DOUBLE || REALTYPE_FLOAT)
```
might be more future-proof. This is only if we consider adding additional types in the future, though.